### PR TITLE
[Sparse] SIGN example.

### PR DESCRIPTION
**Pull request by frozenbugs**
## Checklist
Please feel free to remove inapplicable items for your PR.
- [x] The PR title starts with [$CATEGORY] (such as [NN], [Model], [Doc], [Feature]])
- [x] I've leverage the [tools](https://docs.google.com/document/d/1iHyj7zlmygKSk5gBPsqIqL5ASPzJSPREaNT_QdsiYA4/edit) to beautify the python and c++ code.
- [x] The PR is complete and small, read the [Google eng practice (CL equals to PR)](https://google.github.io/eng-practices/review/developer/small-cls.html) to understand more about small PR. In DGL, we consider PRs with less than 200 lines of core code change are small (example, test and documentation could be exempted).
- [x] All changes have test coverage
- [x] Code is well-documented
- [x] To the best of my knowledge, examples are either not affected by this change, or have been fixed to be compatible with this change
- [x] Related issue is referred in this PR
- [x] If the PR is for a new model/paper, I've updated the example index [here](../examples/README.md).

### CODE REVIEW

File: examples/sparse/sign.py [hunk 1]
```diff
@@ -0,0 +1,112 @@
...
+        # Compute accuracy on training/validation/test.
+        train_acc = (pred[train_mask] == labels[train_mask]).float().mean()
+        val_acc = (pred[val_mask] == labels[val_mask]).float().mean()
+        test_acc = (pred[test_mask] == labels[test_mask]).float().mean()
+
+        # Save the best validation accuracy and the corresponding test
+        # accuracy.
+        if best_val_acc < val_acc:
+            best_val_acc = val_acc
+            best_test_acc = test_acc
```
Comment (mufeili) [inline examples/sparse/sign.py:100]: I don't think it's necessary to have L98-100 for a demonstration.

File: examples/sparse/sign.py [hunk 2]
```diff
@@ -0,0 +1,112 @@
...
+    train_mask = g.ndata["train_mask"]
+    val_mask = g.ndata["val_mask"]
+    test_mask = g.ndata["test_mask"]
+
+    loss_func = nn.CrossEntropyLoss()
+    optimizer = Adam(model.parameters(), lr=3e-3)
+    best_val_acc = 0
+    best_test_acc = 0
+
+    for e in range(100):
```
Comment (mufeili) [inline examples/sparse/sign.py:81]: 10 epochs should be sufficient for this experiment.

File: examples/sparse/sign.py [hunk 3]
```diff
@@ -0,0 +1,112 @@
...
+        if best_val_acc < val_acc:
+            best_val_acc = val_acc
+            best_test_acc = test_acc
+
+        # Backward.
+        optimizer.zero_grad()
+        loss.backward()
+        optimizer.step()
+
+        if e % 5 == 0:
```
Comment (mufeili) [inline examples/sparse/sign.py:107]: `e % 5 == 0` is mostly for a slow experiment and may not be necessary for this one.

File: examples/sparse/sign.py [hunk 4]
```diff
@@ -0,0 +1,112 @@
...
+
+    # 2-hops diffusion operators.
+    r = 2
+    X = g.ndata["feat"]
+    X_sign = sign_diffusion(A, X, r)
+
+    # Create SIGN model.
+    in_size = X.shape[1]
+    out_size = dataset.num_classes
+    model = Sign(in_size, out_size, r)
```
Comment (jermainewang) [inline examples/sparse/sign.py:69]: Maybe call it `SIGN` instead of `Sign` to align with the model name?

**Comment by jermainewang** [comment]
The example looks fine to me. One nit question. According to https://github.com/dmlc/dgl/blob/master/examples/pytorch/graphsage/node_classification.py , there are usually a `train` function and a `evaluate` function instead of a long `__main__` function. Which practice shall we stick with? I'm fine with both but be aware of the refactor effort going forward if we want to make all examples consistent.

File: examples/sparse/sign.py [hunk 5]
```diff
@@ -0,0 +1,111 @@
+"""
+[SIGN: Scalable Inception Graph Neural Networks]
+(https://arxiv.org/abs/2004.11198)
+
+This example shows a simplified version of SIGN: a precomputed 2-hops diffusion
+operators on top of symmetrically normalized adjacency matrix A.
```
Comment (mufeili) [inline examples/sparse/sign.py:6]: "a precomputed 2-hops diffusion operators" grammar error

File: examples/sparse/sign.py [hunk 6]
```diff
@@ -0,0 +1,111 @@
...
+import torch
+import torch.nn as nn
+import torch.nn.functional as F
+from torch.optim import Adam
+
+from dgl.data import CoraGraphDataset
+from dgl.mock_sparse import create_from_coo, diag, identity
+
+################################################################################
+# (HIGHLIGHT) Take the advantage of DGL sparse API to significantly simplify the
```
Comment (mufeili) [inline examples/sparse/sign.py:18]:
```suggestion
# (HIGHLIGHT) Take the advantage of DGL sparse APIs to significantly simplify the
```

Comment (mufeili) [inline examples/sparse/sign.py:18]: simplify against what?

File: examples/sparse/sign.py [hunk 7]
```diff
@@ -0,0 +1,111 @@
...
+# code for sign diffusion operator.
```
Comment (mufeili) [inline examples/sparse/sign.py:19]:
```suggestion
# code for feature diffusion in SIGN
```
?

File: examples/sparse/sign.py [hunk 8]
```diff
@@ -0,0 +1,111 @@
...
+################################################################################
+def sign_diffusion(A, X, r):
+    # Perform the r-hops diffusion operation.
```
Comment (mufeili) [inline examples/sparse/sign.py:22]:
```suggestion
    # Perform the r-hop diffusion operation.
```
no s with hyphen

File: examples/sparse/sign.py [hunk 9]
```diff
@@ -0,0 +1,111 @@
...
+    return X_sign
+
+
+class SIGN(nn.Module):
+    def __init__(self, in_size, out_size, r, hidden_size=64):
+        super().__init__()
+        self.theta = nn.ModuleList(
+            [nn.Linear(in_size, hidden_size) for _ in range(r + 1)]
+        )
+        self.omega = nn.Linear(hidden_size * (r + 1), out_size)
```
Comment (mufeili) [inline examples/sparse/sign.py:36]: I never see people use `theta` and `omega` in defining a model class. If you really want to do that, leave a comment for the equation.

File: examples/sparse/sign.py [hunk 10]
```diff
@@ -0,0 +1,111 @@
...
+
+    def forward(self, X_sign):
+        results = []
+        for i in range(len(X_sign)):
+            results.append(self.theta[i](X_sign[i]))
+        Z = F.relu(torch.cat(results, dim=1))
+        return F.sigmoid(self.omega(Z))
```
Comment (mufeili) [inline examples/sparse/sign.py:43]: Why do you use `F.sigmoid` for multi-class classification?

File: examples/sparse/sign.py [hunk 11]
```diff
@@ -0,0 +1,111 @@
...
+    test_acc = (pred[test_mask] == labels[test_mask]).float().mean()
+    return val_acc, test_acc
+
+
+def train(g, model):
+    labels = g.ndata["label"]
+    train_mask = g.ndata["train_mask"]
+    optimizer = Adam(model.parameters(), lr=3e-3)
+
+    for e in range(20):
```
Comment (mufeili) [inline examples/sparse/sign.py:62]: Use `epoch` which is more semantically meaningful.

File: examples/sparse/sign.py [hunk 12]
```diff
@@ -0,0 +1,111 @@
...
+        # Evaluate the prediction.
+        val_acc, test_acc = evaluate(g, pred)
+        print(
+            f"In epoch {e}, loss: {loss:.3f}, val acc: {val_acc:.3f}, test acc"
+            f": {test_acc:.3f}"
+        )
+
+
+if __name__ == "__main__":
+    # If CUDA is available, use GPU to accelerate the training, otherwise, use
```
Comment (mufeili) [inline examples/sparse/sign.py:86]:
```suggestion
    # If CUDA is available, use GPU to accelerate the training, use CPU otherwise
```
to save L87

File: examples/sparse/sign.py [hunk 13]
```diff
@@ -0,0 +1,111 @@
...
+    # CPU for the training.
+    dev = torch.device("cuda:0" if torch.cuda.is_available() else "cpu")
+
+    # Create graph from the existing dataset.
+    dataset = CoraGraphDataset()
+    g = dataset[0].to(dev)
+
+    # Create the sparse adjacent matrix A (note that W was used as notion for
```
Comment (mufeili) [inline examples/sparse/sign.py:94]:
```suggestion
    # Create the sparse adjacency matrix A (note that W was used as the notation for
```

File: examples/sparse/sign.py [hunk 14]
```diff
@@ -0,0 +1,111 @@
...
+    # adjacent matrix in the original paper).
```
Comment (mufeili) [inline examples/sparse/sign.py:95]:
```suggestion
    # the adjacency matrix in the original paper).
```

File: examples/sparse/sign.py [hunk 15]
```diff
@@ -0,0 +1,111 @@
...
+        # Backward.
+        optimizer.zero_grad()
+        loss.backward()
+        optimizer.step()
+
+        # Evaluate the prediction.
+        val_acc, test_acc = evaluate(g, pred)
+        print(
+            f"In epoch {e}, loss: {loss:.3f}, val acc: {val_acc:.3f}, test acc"
+            f": {test_acc:.3f}"
```
Comment (mufeili) [inline examples/sparse/sign.py:81]: It's a bit controversial to print test accuracy every epoch as some people consider it as cheating.

**Review by mufeili — CHANGES_REQUESTED** [review]
Done a pass

File: examples/sparse/sign.py [hunk 16]
```diff
@@ -0,0 +1,111 @@
...
+    labels = g.ndata["label"]
+    train_mask = g.ndata["train_mask"]
+    optimizer = Adam(model.parameters(), lr=3e-3)
+
+    for e in range(20):
+        # Forward.
+        logits = model(X_sign)
+
+        # Compute prediction.
+        pred = logits.argmax(1)
```
Comment (mufeili) [inline examples/sparse/sign.py:67]: Place this line before L78.

File: examples/sparse/sign.py (hunk 15 above)
Comment (frozenbugs) [reply examples/sparse/sign.py]: Since this is an example, it should be fine to show it.

**Comment by frozenbugs** [comment]
> The example looks fine to me. One nit question. According to https://github.com/dmlc/dgl/blob/master/examples/pytorch/graphsage/node_classification.py , there are usually a `train` function and a `evaluate` function instead of a long `__main__` function. Which practice shall we stick with? I'm fine with both but be aware of the refactor effort going forward if we want to make all examples consistent.

Updated.

**Review by jermainewang — APPROVED** [review]

**Review by mufeili — APPROVED** [review]
